lib/native/http.ts

```diff
@@ -18,31 +18,69 @@ import { NativeResource } from "./native_resource";
 import { ResourceSafe } from '../common/resource_safety';
 import { ClientBootstrap, ClientTlsContext, SocketOptions } from './io';
 
-type ConnectionCallback = (error_code: Number) => void;
-
 export class HttpConnection extends NativeResource implements ResourceSafe {
 
-    constructor(
+    protected constructor(native_handle: any)
+    {
+        super(native_handle);
+    }
+
+    close() {
+        crt_native.http_connection_close(this.native_handle());
+    }
```

Comment:
You're never calling `aws_http_connection_release()` anywhere.

`aws_http_connection_close()` tells the connection to shutdown the channel, if it's not already shut down, but this won't necessarily clean up the `aws_http_connection` struct.

The connection isn't cleaned up until BOTH `aws_http_connection_release()` is called AND the on_shutdown callback has fired.

If this `close()` function is supposed to mean "finalize resource" then you probably want it to call `aws_http_connection_release()` instead. But you'll still need to keep the JS object alive until on_shutdown has fired (unless it's fired already). 

```
/**
 * Users must release the connection when they are done with it.
 * The connection's memory cannot be reclaimed until this is done.
 * If the connection was not already shutting down, it will be shut down.
 *
 * Users should always wait for the on_shutdown() callback to be called before releasing any data passed to the
 * http_connection (Eg aws_tls_connection_options, aws_socket_options) otherwise there will be race conditions between
 * http_connection shutdown tasks and memory release tasks, causing Segfaults.
 */
AWS_HTTP_API
void aws_http_connection_release(struct aws_http_connection *connection);
```

Reply:
Yeah, I haven't been able to test this yet, and didn't get there yet. This was mostly so you could see the ts code. I'll get it sorted though.